src/ConfigManager.php

```diff
@@ -537,16 +559,32 @@ public function isFinancialProductAvailable($product_type, array $products)
 
         if (isset($product_category_filters[$product_type]) && count($product_category_filters[$product_type])) {
             $excluded_cat_ids = $product_category_filters[$product_type];
+            $available_cat_ids = array_diff(array_keys($this->getAllProductCategories()), $excluded_cat_ids);
+
+            $parent_categories = [];
 
             foreach ($products as $product) {
                 $category_id = (int) $product['id_category_default'];
 
-                if (in_array($category_id, $excluded_cat_ids, true)
-                    || count(array_intersect($excluded_cat_ids, array_map(
-                        static function (\Category $category) { return $category->id; },
-                        (new \Category($category_id))->getAllChildren()->getResults()
-                    )))
-                ) {
+                if (in_array($category_id, $excluded_cat_ids, true)) {
+                    foreach (array_diff($available_cat_ids, [$category_id]) as $cat_id) {
+                        if (!isset($parent_categories[$cat_id])) {
+                            $parent_categories[$cat_id] = [];
+
+                            if (is_array($cat_parents = (new \Category($cat_id))->getParentsCategories())) {
+                                foreach ($cat_parents as $category) {
+                                    if ($category['id_category'] !== $cat_id) {
+                                        $parent_categories[$cat_id][] = $category['id_category'];
+                                    }
+                                }
+                            }
+                        }
+
+                        if (in_array($category_id, $parent_categories[$cat_id], true)) {
+                            continue 2;
+                        }
+                    }
+
```

Comment:
🤯 4 ify 3 foreache, ciężko się czyta. Myślałeś może, żeby to jakoś rozbić?

Reply:
A jak to niby rozbić? Nie zaglądaj do źródeł systemów operacyjnych albo silników RDBMS. Tam to dopiero można zobaczyć zagnieżdżenia. ;-)

Reply:
Co najwyżej możnasprawdzanie parenta wyrzucić do funkcji pomocniczej, ale to już jak będzie kompletny refaktor i wydzielanie kodu na zewnątrz w jedno miejsce. W WC nie miałem tego problemu, bo jest gotowa funkcja, czy dana kategoria jest rodzicem dla innej wskazanej i koda w wersji dla WC jest dużo prostszy wizualnie, mimo, że robi to samo. Niestety w PS trzeba było więcej samemu zaimplementować, bo nie znalazłem odpowiednika funkcji term_is_ancestor_of z WC, chociaż tam dla odmiany jest inna logika sprawdzajaca w jakim formacie dostaję dane produktu (raz jest tablica, raz obiekt WC_Product_Simple).

Reply:
To wszystko będzie zaorane i przepisane jednolicie na nowo w jednej generycznej wersji w blibliotece z kodem biznesowym, obok bilbioteki z kodem do komunikacji z API (ta biblioteka już jest, ale nieprzetestowana).